_posts/2024-09-05-digging-into-struct-initialization-performance.md

```diff
@@ -0,0 +1,251 @@
+---
+title: "Digging into struct initialization performance"
+author: ysbaddaden
+summary: "How a quick test of BLAKE3 vs SHA256 in Crystal unveiled a performance issue that stems from the Ruby syntax."
+categories: technical
+tags: [language,performance]
+---
+
+I toyed with BLAKE3 some months ago. I wanted to verify how much faster was BLAKE3 versus SHA256 in Crystal, and see if it would bring some free improvement. Luckily there is [a shard](https://github.com/didactic-drunk/blake3.cr) wrapping the [official library](https://github.com/BLAKE3-team/BLAKE3) that is highly optimized with custom assembly for different CPU features (SSE, AVX2, NEON …).
+
+To my surprise, performance was pretty bad, and this post is the written legacy of the quest to understand why. Be prepared to dig into technical details and how language design can impact performance.
+
+## Let’s benchmark
+
+One of the points for BLAKE3 is that it’s magnitudes faster than alternative hash digests; the authors claim almost 14 times faster than SHA256 for example. I wrote a quick benchmark, comparing `Digest::Blake3` to `Digest::SHA256` (backed by OpenSSL) that is usually my default choice for my use cases, for example to hash a session id made of 32 bytes.
+
+```crystal
```

Comment:
I feel like that would be too much. Maybe there could be a gist and we could link to it (but I think I lack some courage).

Reply:
It's not much:
````md
```yml
name: blake-struct-test
version: 0.1.0

dependencies:
  blake3:
    github: didactic-drunk/blake3.cr
```
````

It would be really helpful for people who want to follow along at home, which I think is an important exercise with technical posts like this.